src/box/iproto.cc

```diff
@@ -3519,6 +3538,48 @@ iproto_override_event_init(struct event *event, void *arg)
 	return true;
 }
 
+/**
+ * Notifies IPROTO threads that a new request handler has been set.
+ */
+static void
+iproto_cfg_override(uint32_t req_type, bool is_set);
+
+/**
+ * Trigger which is fired on any change in the event registry.
+ */
+static int
+trigger_on_change_iproto_notify(struct trigger *trigger, void *arg)
+{
+	(void)trigger;
+	bool is_by_id;
+	struct event *event = (struct event *)arg;
+	uint32_t type = get_iproto_type_from_event_name(event->name, &is_by_id);
+	if (type == iproto_type_MAX)
+		return 0;
+
+	struct iproto_req_handlers *handlers;
+	handlers = iproto_req_handlers_get(type);
+	bool old_is_set = iproto_req_handler_is_set(handlers);
+
+	if (event_has_triggers(event)) {
+		if (handlers == NULL)
+			handlers = iproto_req_handlers_new(type);
+		iproto_req_handlers_set_event(handlers, event, is_by_id);
+	} else {
+		iproto_req_handlers_del_event(handlers, is_by_id);
+	}
+
+	bool new_is_set = iproto_req_handler_is_set(handlers);
+	if (!new_is_set)
+		iproto_req_handlers_delete(handlers, type);
```

Comment:
Looks like a bug: you can delete a handler from TX only after notifying IPROTO. Otherwise there can be inflight requests that arrive at the override route only to see there is no request handler set (which is unexpected).

Reply:
BTW, please also refactor the common logic out into a separate helper.

Reply:
Thanks, fixed.

> BTW, please also refactor the common logic out into a separate helper.

I tried, but couldn't make the code more beautiful than it is now.

Reply:
I mean these two snippets are very similar, I suggest refactoring them into a separate function:

https://github.com/Gumix/tarantool/blob/f5953480a14118dd06efb8f5668b398afc5f4637/src/box/iproto.cc#L3653-L3660

https://github.com/Gumix/tarantool/blob/f5953480a14118dd06efb8f5668b398afc5f4637/src/box/iproto.cc#L4322-L4329

Feel free to ignore.

Reply:
Moved to the new function `iproto_override_finish()`.